// undux/dist/src/react.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var utils_1 = require("./utils");
function connect(store) {
    return function (Component) {
        return _a = /** @class */ (function (_super) {
                __extends(class_1, _super);
                function class_1(props) {
                    var _this = _super.call(this, props) || this;
                    var lasts = {};
                    _this.state = {
                        store: store['store'],
                        subscriptions: [
                            store.beforeAll().subscribe(function (_a) {
                                var key = _a.key, previousValue = _a.previousValue;
                                lasts[key] = previousValue;
                            }),
                            store.onAll().subscribe(function (_a) {
                                var key = _a.key, value = _a.value;
                                if (lasts[key] && utils_1.equals(lasts[key], value)) {
                                    return false;
                                }
                                _this.setState({ store: store['store'] });
                            })
                        ]
                    };
                    return _this;
                }
                class_1.prototype.componentWillUnmount = function () {
                    this.state.subscriptions.forEach(function (_) { return _.unsubscribe(); });
                };
                class_1.prototype.shouldComponentUpdate = function (props, state) {
                    var _this = this;
                    return state.store !== this.state.store
                        || Object.keys(props).some(function (_) { return props[_] !== _this.props[_]; });
                };
                class_1.prototype.render = function () {
                    return React.createElement(Component, __assign({}, this.props, { store: this.state.store }));
                };
                return class_1;
            }(React.Component)),
            _a.displayName = "withStore(" + utils_1.getDisplayName(Component) + ")",
            _a;
        var _a;
    };
}
exports.connect = connect;
